src/generateUrls.js

```diff
@@ -50,20 +69,56 @@ function generateUrls(router, options) {
       }
     }
 
-    let path = '';
-    while (route) {
-      if (route.path !== '/') {
-        let toPath = cache.get(route.path);
-        if (!toPath) {
-          toPath = Router.pathToRegexp.compile(route.path);
-          cache.set(route.path, toPath);
+    let regexp = cache.get(route.fullPath);
+    if (!regexp) {
+      let fullPath = '';
+      while (route) {
+        if (route.path !== '/') {
+          fullPath = route.path + fullPath;
+        }
+        route = route.parent;
+      }
+      const tokens = Router.pathToRegexp.parse(fullPath);
+      const toPath = Router.pathToRegexp.tokensToFunction(tokens);
+      const keys = Object.create(null);
+      for (let i = 0; i < tokens.length; i += 1) {
+        if (typeof tokens[i] !== 'string') {
+          keys[tokens[i].name] = true;
         }
-        path = toPath(params, options) + path;
       }
-      route = route.parent;
+      regexp = { toPath, keys };
+      cache.set(fullPath, regexp);
+      router.routesByName[routeName].fullPath = fullPath;
+    }
+
+    let url = router.baseUrl + regexp.toPath(params, options) || '/';
+    if (options.query && params) {
+      const encode = options.pretty ? encodeURIComponentPretty : encodeURIComponent;
+      const keys = Object.keys(params);
+      const pairs = [];
+
+      for (let i = 0; i < keys.length; i += 1) {
+        const key = keys[i];
+        if (!regexp.keys[key]) {
+          const ks = `${encode(stringifyPrimitive(key))}=`;
+          const val = params[key];
+
+          if (Array.isArray(val)) {
+            for (let j = 0; j < val.length; j += 1) {
+              pairs.push(ks + encode(stringifyPrimitive(val[j])));
+            }
+          } else {
+            pairs.push(ks + encode(stringifyPrimitive(val)));
+          }
+        }
+      }
```

Comment:
There are many ways how users will want to stringify query params. So it would be nice to give them a possibility to use something as [qs](https://www.npmjs.com/package/qs).

May be, we can re-use `query` option. "boolean or function that will stringify the params".

So, from this perspective, this code should be splitted into two parts:
1. build a query params object.
2. use built-in or custom stringify function

May be, it's better to do not provide built-in stringify function at all. Do you really want to support this code?